master/buildbot/test/integration/test_latent.py

```diff
@@ -0,0 +1,224 @@
+        builder_id = self.successResultOf(getBuilderIdByName(master, 'testy'))
+
+        # Trigger a buildrequest
+        bsid, brids = self.successResultOf(
+            master.data.updates.addBuildset(
+                waited_for=False,
+                builderids=[builder_id],
+                sourcestamps=[
+                    {'codebase': '',
+                     'repository': '',
+                     'branch': None,
+                     'revision': None,
+                     'project': ''},
+                ],
+            )
+        )
+
+        unclaimed_build_requests = []
+        self.successResultOf(master.mq.startConsuming(
+            lambda key, request: unclaimed_build_requests.append(request),
+            ('buildrequests', None, 'unclaimed')))
+
+        # The worker fails to substantiate.
+        controller.start_instance(Failure(TestException("substantiation failed")))
+        # Flush the errors logged by the failure.
+        self.flushLoggedErrors(TestException)
+
+        # When the substantiation fails, the buildrequest becomes unclaimed.
+        self.assertEqual(
+            set(brids),
+            {req['buildrequestid'] for req in unclaimed_build_requests},
+        )
+
+    def test_worker_accepts_builds_after_failure(self):
+        """
+        If a latent worker fails to substantiate, the worker is still able to accept jobs.
+        """
+        controller = LatentController('local')
+        config_dict = {
+            'builders': [
+                BuilderConfig(name="testy",
+                              workernames=["local"],
+                              factory=BuildFactory(),
+                              ),
+            ],
+            'workers': [controller.worker],
+            'protocols': {'null': {}},
+            'multiMaster': True,
+        }
+        master = self.successResultOf(getMaster(self, self.reactor, config_dict))
+        builder_id = self.successResultOf(getBuilderIdByName(master, 'testy'))
+
+        # Trigger a buildrequest
+        bsid, brids = self.successResultOf(
+            master.data.updates.addBuildset(
+                waited_for=False,
+                builderids=[builder_id],
+                sourcestamps=[
+                    {'codebase': '',
+                     'repository': '',
+                     'branch': None,
+                     'revision': None,
+                     'project': ''},
+                ],
+            )
+        )
+
+        unclaimed_build_requests = []
+        self.successResultOf(master.mq.startConsuming(
+            lambda key, request: unclaimed_build_requests.append(request),
+            ('buildrequests', None, 'unclaimed')))
+
+        # The worker fails to substantiate.
+        controller.start_instance(Failure(TestException("substantiation failed")))
+        # Flush the errors logged by the failure.
+        self.flushLoggedErrors(TestException)
+
+        self.assertEqual(controller.started, True)
```

Comment:
The issue that this catches (and you can verify that it is an issue if you run the test without the corresponding code changes), is that worker doesn't appear as latent but instead as disconnected after failing to substantiate. And so nothing will try to start an instance ever again.

I've added a comment explaining why I test this here.